lib/floe/workflow/choice_rule/data.rb

```diff
@@ -4,88 +4,36 @@ module Floe
-          when "StringEquals", "StringEqualsPath",
-               "NumericEquals", "NumericEqualsPath",
-               "BooleanEquals", "BooleanEqualsPath",
-               "TimestampEquals", "TimestampEqualsPath"
-            lhs == rhs
-          when "StringLessThan", "StringLessThanPath",
-               "NumericLessThan", "NumericLessThanPath",
-               "TimestampLessThan", "TimestampLessThanPath"
-            lhs < rhs
-          when "StringGreaterThan", "StringGreaterThanPath",
-               "NumericGreaterThan", "NumericGreaterThanPath",
-               "TimestampGreaterThan", "TimestampGreaterThanPath"
-            lhs > rhs
-          when "StringLessThanEquals", "StringLessThanEqualsPath",
-               "NumericLessThanEquals", "NumericLessThanEqualsPath",
-               "TimestampLessThanEquals", "TimestampLessThanEqualsPath"
-            lhs <= rhs
-          when "StringGreaterThanEquals", "StringGreaterThanEqualsPath",
-               "NumericGreaterThanEquals", "NumericGreaterThanEqualsPath",
-               "TimestampGreaterThanEquals", "TimestampGreaterThanEqualsPath"
-            lhs >= rhs
-          when "StringMatches"
-            lhs.match?(Regexp.escape(rhs).gsub('\*', '.*?'))
-          else
-            raise Floe::InvalidWorkflowError, "Invalid choice [#{compare_key}]"
+        DATA_RULES = {
+          "IsNull"            => Floe::Workflow::ChoiceRule::IsNull,
+          "IsPresent"         => Floe::Workflow::ChoiceRule::IsPresent,
+          "IsNumeric"         => Floe::Workflow::ChoiceRule::IsNumeric,
+          "IsString"          => Floe::Workflow::ChoiceRule::IsString,
+          "IsBoolean"         => Floe::Workflow::ChoiceRule::IsBoolean,
+          "IsTimestamp"       => Floe::Workflow::ChoiceRule::IsTimestamp,
+          "Equals"            => Floe::Workflow::ChoiceRule::Equals,
+          "LessThan"          => Floe::Workflow::ChoiceRule::LessThan,
+          "GreaterThan"       => Floe::Workflow::ChoiceRule::GreaterThan,
+          "LessThanEquals"    => Floe::Workflow::ChoiceRule::LessThanEquals,
+          "GreaterThanEquals" => Floe::Workflow::ChoiceRule::GreaterThanEquals,
+          "Matches"           => Floe::Workflow::ChoiceRule::Matches
+        }.freeze
+        COMPARE_RULE = /^(String|Numeric|Boolean|Timestamp)?(#{DATA_RULES.keys.join("|")})(Path)?$/.freeze
+
+        def self.build(payload)
+          compare_key, klass = klass_params(payload)
+          raise Floe::InvalidWorkflowError, "Data-test Expression Choice Rule must have a compare key" unless compare_key
+          raise Floe::InvalidWorkflowError, "Data-test Expression Choice Rule must have a valid compare key" unless klass
+
+          klass.new(payload, :compare_key => compare_key)
+        end
+
+        def self.klass_params(payload)
+          payload.each_key do |key|
+            values = COMPARE_RULE.match(key)
+            return [key, DATA_RULES[values[2]], !!values[3]] if values
```

Comment:
I think this would read better if the match was decomposed or perhaps if named captures were used instead of using values[2], values[3], etc.

Some other quetsions
- where is the prefix used (the `(String|Numeric|Boolean|Timestamp)` part)?  It seems to be ignored here.
- The caller does `compare_key, klass = klass_params(payload)`, but this method returns 3 items, so is that a bug?  Where did the 3rd param go with the path, and why isn't it being used?

Reply:
the current pattern for the states/command objects to parse the payload themselves.
Here, we need to parse the string to 

The old code ignored the first part, so I continued to ignore it. I can see doing type checks or conversions.

returning 2 vs 3 is probably a bug. thanks

Reply:
1. yes, we ignore the type prefix part and don't currently do data type conversion. (not sure if that is needed)
2. changed this code. think all set.